### DISCUSSION TITLE
How does CVXPY obtain primal variables after solving the dual problem for conic problems?

### DISCUSSION THREAD

**Discussion by nielsdekoeijer** [cvxpy/cvxpy#1699] [Q&A] [answered]
After prototyping a problem using CVXPY, I am trying to reverse-engineer what it is doing.
I am solving a conic problem and the MOSEK solver is being invoked.

It is my understanding that CVXPY first converts the problem to its internal conic representation given by:
`(P-Opt) min{ c.T @ x : A @ x + b in K } + d`

However, before passing it to MOSEK, CVXPY seems to convert the problem to the dual form given by:
`(D-Opt) max{ -b @ y : c = A.T @ y, y in K^* } + d`

My question is how the "primal variable" `x` is then retrieved from the "dual variable" `y`.

I have created my own implementation, and I have (probably naively) attempted to solve the system:
`A @ x + b = y`
The resulting `x` is similar to the result of solving `(P-Opt)` directly, but slightly worse (in terms of primal cost).

**Comment by rileyjmurray** [comment, marked as answer]
We use the fact that the dual of the dual is the primal (generically). You
can maybe see how the dualize reduction works by looking at its unit tests.
We have a tests somewhere with a simple function that simulates the dualize
reduction.

**Reply by nielsdekoeijer** [reply]
Thanks for the quick response! Unit tests were illuminating.

For those in a similar situation as myself: I was able to recover the original `x` of `(P-opt)` by investigating the dual variables of the equality constraint in `D-opt`.
